File: stage-common/src/main/java/io/github/solusmods/eternalcore/stage/impl/network/c2s/ClientAccess.java
package io.github.solusmods.eternalcore.stage.impl.network.c2s;

import io.github.solusmods.eternalcore.stage.api.Stage;
import io.github.solusmods.eternalcore.stage.api.StageAPI;
import io.github.solusmods.eternalcore.stage.api.StageInstance;
import io.github.solusmods.eternalcore.stage.api.Stages;
import io.github.solusmods.eternalcore.stage.impl.StageStorage;
import lombok.AccessLevel;
import lombok.NoArgsConstructor;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.entity.player.Player;

import java.util.Optional;

@NoArgsConstructor(access = AccessLevel.PRIVATE)
public class ClientAccess {

    static void handle(SyncStagesStoragePayload packet, ServerPlayer player) {
        player.eternalCore$getStorageOptional(StageStorage.getKey()).
                ifPresent(storage -> storage.load(packet.data()));
    }

    static void handle(RequestStageBreakthroughPacket packet, Player player){
        if (player == null) return;

        Stages storage = StageAPI.getStageFrom(player);
        Optional<StageInstance> optional = storage.getStage();
        if (optional.isEmpty()) return;

        Stage stage = StageAPI.getStageRegistry().get(packet.stage());
        if (stage == null) return;

        StageInstance instance = optional.get();
        if (!instance.getNextBreakthroughs(player).contains(stage)) {
        }

        storage.breakthroughStage(stage);
    }
}

File: stage-common/src/main/kotlin/io/github/solusmods/eternalcore/stage/impl/network/c2s/ClientAccess.kt
package io.github.solusmods.eternalcore.stage.impl.network.c2s

import io.github.solusmods.eternalcore.stage.api.StageAPI
import io.github.solusmods.eternalcore.stage.impl.StageStorage
import lombok.AccessLevel
import lombok.NoArgsConstructor
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.entity.player.Player
import java.util.function.Consumer

@NoArgsConstructor(access = AccessLevel.PRIVATE)
object ClientAccess {
    fun handle(packet: SyncStagesStoragePayload, player: ServerPlayer) {
        player.`eternalCore$getStorageOptional`(StageStorage.key!!).ifPresent { storage: StageStorage ->
            storage.load(
                packet.data!!
            )
        }
    }

    fun handle(packet: RequestStageBreakthroughPacket, player: Player) {

        val storage = StageAPI.getStageFrom(player)
        val optional = storage!!.getStageOptional()
        if (optional.isEmpty) return

        val stage = StageAPI.stageRegistry.get(packet.stage)
        if (stage == null) return

        val instance = optional.get()
        if (!instance.getNextBreakthroughs(player)!!.contains(stage)) {
        }

        storage.breakthroughStage(stage)
    }
}
